[JSC] Optimize Function.prototype.bind

Constellation opened the pull request:
#### 1c992e3ed9b1a3c81b0dcf7bec31089caf0b9fbb
<pre>
[JSC] Optimize Function.prototype.bind
<a href="https://bugs.webkit.org/show_bug.cgi?id=254054">https://bugs.webkit.org/show_bug.cgi?id=254054</a>
rdar://problem/106837856

Reviewed by Mark Lam.

This patch improves Function.prototype.bind a bit. We are planning to improve this further
to avoid arguments buffer for small # of arguments cases in a subsequent patch. This changes
two things.

1. Make canAssumeNameAndLengthAreOriginal work for bound functions. It helps when we have bind().bind() calls,
   which can be observed in Speedometer2.
2. Make everything in C++ since we already have a mechanism to avoid costly operation in C++. So we can implement
   Function#bind super efficiently in C++ at this point, let&apos;s just make it back.

It improves microbenchmark results.

                                                  ToT                     Patched

function-bind-creation                      15.1042+-0.0587     ^     12.6237+-0.0534        ^ definitely 1.1965x faster
function-bind-no-inlining-repeat-call       43.6672+-0.2105     ^     42.4584+-0.2360        ^ definitely 1.0285x faster
function-bind                              165.5037+-0.7954     ^     61.0135+-0.1911        ^ definitely 2.7126x faster

* JSTests/microbenchmarks/function-bind.js: Added.
(test):
* Source/JavaScriptCore/builtins/BuiltinNames.h:
* Source/JavaScriptCore/builtins/FunctionPrototype.js:
(overriddenName.string_appeared_here.symbolHasInstance):
(bind): Deleted.
* Source/JavaScriptCore/bytecode/LinkTimeConstant.h:
* Source/JavaScriptCore/runtime/FunctionPrototype.cpp:
(JSC::FunctionPrototype::addFunctionProperties):
(JSC::JSC_DEFINE_HOST_FUNCTION):
* Source/JavaScriptCore/runtime/FunctionRareData.cpp:
(JSC::FunctionRareData::FunctionRareData):
* Source/JavaScriptCore/runtime/FunctionRareData.h:
* Source/JavaScriptCore/runtime/JSFunction.cpp:
(JSC::JSFunction::put):
(JSC::JSFunction::deleteProperty):
(JSC::JSFunction::defineOwnProperty):
(JSC::JSFunction::reifyLength):
(JSC::makeNameWithOutOfMemoryCheck): Deleted.
* Source/JavaScriptCore/runtime/JSFunction.h:
* Source/JavaScriptCore/runtime/JSFunctionInlines.h:
(JSC::JSFunction::originalLength):
(JSC::makeNameWithOutOfMemoryCheck):
(JSC::JSFunction::originalName):
(JSC::JSFunction::canAssumeNameAndLengthAreOriginal):
* Source/JavaScriptCore/runtime/JSGlobalObject.cpp:
(JSC::JSGlobalObject::init):

Canonical link: <a href="https://commits.webkit.org/261825@main">https://commits.webkit.org/261825@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f3cea9ea3660002348486c07287be64560f53d4b

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/112994 "Passed style check") | [  ~~🛠 ios~~](https://ews-build.webkit.org/#/builders/77/builds/22143 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [❌ 🛠 mac](https://ews-build.webkit.org/#/builders/90/builds/1668 "Failed to checkout and rebase branch from PR 11632") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/87/builds/4767 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/121479 "Built successfully") 
| [  ~~🧪 bindings~~](https://ews-build.webkit.org/#/builders/11/builds/117102 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/23500 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/13314 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/86/builds/5978 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/118781 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/17458 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [❌ 🧪 api-mac](https://ews-build.webkit.org/#/builders/90/builds/1668 "Failed to checkout and rebase branch from PR 11632") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/36/builds/106084 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/99416 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [❌ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/90/builds/1668 "Failed to checkout and rebase branch from PR 11632") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/106084 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [  ~~🛠 🧪 jsc~~](https://ews-build.webkit.org/#/builders/1/builds/101250 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/14442 "Built successfully") | [❌ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/90/builds/1668 "Failed to checkout and rebase branch from PR 11632") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/36/builds/106084 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [  ~~🛠 🧪 jsc-arm64~~](https://ews-build.webkit.org/#/builders/83/builds/12598 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/15151 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/84/builds/10631 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 jsc-armv7~~](https://ews-build.webkit.org/#/builders/38/builds/102788 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/20457 "Built successfully") | [❌ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/90/builds/1668 "Failed to checkout and rebase branch from PR 11632") | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/38/builds/102788 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/17001 "Built successfully") | | [  ~~🛠 jsc-mips~~](https://ews-build.webkit.org/#/builders/37/builds/110839 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/4526 "Built successfully and passed tests") | | | [  ~~🧪 jsc-mips-tests~~](https://ews-build.webkit.org/#/builders/45/builds/27365 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
<!--EWS-Status-Bubble-End-->

MenloDorian commented on Source/JavaScriptCore/runtime/FunctionPrototype.cpp:
```diff
@@ -102,6 +103,105 @@ JSC_DEFINE_HOST_FUNCTION(functionProtoFuncToString, (JSGlobalObject* globalObjec
+    unsigned numBoundArgs = 0;
+    JSImmutableButterfly* butterfly = nullptr;
+    if (argumentCount > 1) {
+        numBoundArgs = argumentCount - 1;
```
Is this correct?  `callFrame->argumentCount()` already excludes `this`.

Constellation replied:
This is correct. Since `bind`'s first argument is `|this|`

MenloDorian commented on Source/JavaScriptCore/builtins/FunctionPrototype.js:
```diff
@@ -56,37 +56,3 @@ function symbolHasInstance(value)
-    var numBoundArgs = 0;
-    if (argumentCount > 1) {
-        numBoundArgs = argumentCount - 1;
-        boundArgs = @createArgumentsButterflyExcludingThis(this);
```
Since `@createArgumentsButterflyExcludingThis` is no longer used anywhere, let's also remove the `createArgumentsButterflyExcludingThis` bytecode intrinsic and all the manifestations of it in the JIT tiers.  I basically did a case-insensitive grep of `createArgumentsButterflyExcludingThis` and saw it in many places.

It's also ok to remove this in a subsequent patch since the removal touches a lot of code, and is not directly related to this `bind` optimization.

Constellation replied:
Removed.

MenloDorian commented on Source/JavaScriptCore/runtime/JSFunctionInlines.h:
```diff
@@ -123,18 +123,89 @@ inline bool JSFunction::hasReifiedName() const
+
+    if (jsExecutable()->isGetter()) {
+        name = makeNameWithOutOfMemoryCheck(globalObject, scope, "Getter ", "get ", name);
+        RETURN_IF_EXCEPTION(scope, { });
```
nit: you can just use a single `RETURN_IF_EXCEPTION` after this if statement since both paths lead to the exception check with not other code in between.

Constellation replied:
The non-getter-setter case exists too.

MenloDorian replied:
True.  makes sense.

MenloDorian commented on Source/JavaScriptCore/runtime/FunctionRareData.cpp:
```diff
@@ -76,8 +76,8 @@ FunctionRareData::FunctionRareData(VM& vm, ExecutableBase* executable)
-    , m_hasModifiedLengthForNonHostFunction(false)
-    , m_hasModifiedNameForNonHostFunction(false)
+    , m_hasModifiedLengthForNonNativeHostFunction(false)
+    , m_hasModifiedNameForNonNativeHostFunction(false)
```
Per our offline conversation, the reason you added `Native` to the `NonHostFunction` name is because bound functions are also host functions but they are allowed (i.e. not excluded by this non-host filter) as seen in `canAssumeNameAndLengthAreOriginal`.  I suggest using `ForBoundOrNonHostFunction` instead to be more accurate.

MenloDorian reviewed (approved):
r=me

Constellation committed:
[JSC] Optimize Function.prototype.bind
https://bugs.webkit.org/show_bug.cgi?id=254054
rdar://problem/106837856

Reviewed by Mark Lam.

This patch improves Function.prototype.bind a bit. We are planning to improve this further
to avoid arguments buffer for small # of arguments cases in a subsequent patch. This changes
two things.

1. Make canAssumeNameAndLengthAreOriginal work for bound functions. It helps when we have bind().bind() calls,
   which can be observed in Speedometer2.
2. Make everything in C++ since we already have a mechanism to avoid costly operation in C++. So we can implement
   Function#bind super efficiently in C++ at this point, let's just make it back.

It improves microbenchmark results.

                                                  ToT                     Patched

function-bind-creation                      15.1042+-0.0587     ^     12.6237+-0.0534        ^ definitely 1.1965x faster
function-bind-no-inlining-repeat-call       43.6672+-0.2105     ^     42.4584+-0.2360        ^ definitely 1.0285x faster
function-bind                              165.5037+-0.7954     ^     61.0135+-0.1911        ^ definitely 2.7126x faster

* JSTests/microbenchmarks/function-bind.js: Added.
(test):
* Source/JavaScriptCore/builtins/BuiltinNames.h:
* Source/JavaScriptCore/builtins/FunctionPrototype.js:
(overriddenName.string_appeared_here.symbolHasInstance):
(bind): Deleted.
* Source/JavaScriptCore/bytecode/LinkTimeConstant.h:
* Source/JavaScriptCore/runtime/FunctionPrototype.cpp:
(JSC::FunctionPrototype::addFunctionProperties):
(JSC::JSC_DEFINE_HOST_FUNCTION):
* Source/JavaScriptCore/runtime/FunctionRareData.cpp:
(JSC::FunctionRareData::FunctionRareData):
* Source/JavaScriptCore/runtime/FunctionRareData.h:
* Source/JavaScriptCore/runtime/JSFunction.cpp:
(JSC::JSFunction::put):
(JSC::JSFunction::deleteProperty):
(JSC::JSFunction::defineOwnProperty):
(JSC::JSFunction::reifyLength):
(JSC::makeNameWithOutOfMemoryCheck): Deleted.
* Source/JavaScriptCore/runtime/JSFunction.h:
* Source/JavaScriptCore/runtime/JSFunctionInlines.h:
(JSC::JSFunction::originalLength):
(JSC::makeNameWithOutOfMemoryCheck):
(JSC::JSFunction::originalName):
(JSC::JSFunction::canAssumeNameAndLengthAreOriginal):
* Source/JavaScriptCore/runtime/JSGlobalObject.cpp:
(JSC::JSGlobalObject::init):

Canonical link: https://commits.webkit.org/261825@main
